ansible/roles/test/files/ptftests/inner_hash_test.py

```diff
@@ -0,0 +1,505 @@
+
+        for outer_encap_format in outer_encap_formats:
+            hit_count_map = {}
+            for _ in range(0, self.balancing_test_times*len(self.exp_port_list)):
+                logging.info('Checking {} hash key {}, src_port={}, exp_ports={}, dst_ip={}'\
+                    .format(outer_encap_format, hash_key, src_port, self.exp_port_list, self.outer_dst_ip))
+
+                ip_src = self.inner_src_ip_interval.get_random_ip() if hash_key == 'src-ip' else self.inner_src_ip_interval.get_first_ip()
+                ip_dst = self.inner_dst_ip_interval.get_random_ip() if hash_key == 'dst-ip' else self.inner_dst_ip_interval.get_first_ip()
+                sport = random.randint(0, 65535) if hash_key == 'src-port' else 1234
+                dport = random.randint(0, 65535) if hash_key == 'dst-port' else 80
+                ip_proto = self._get_ip_proto() if hash_key == 'ip-proto' else 6
+
+                (packet_port_index, _) = self.check_ip_route(hash_key, outer_encap_format, src_port, ip_src, ip_dst, sport, dport, ip_proto)
+                if self.symmetric_hashing and hash_key != 'ip-proto':
+                    # Send the same packet with reversed tuples and validate that it lands on the same port
+                    rand_src_port = int(random.choice(self.src_ports))
+                    (rPacket_port_index, _) = self.check_ip_route(hash_key, outer_encap_format, src_port, ip_dst, ip_src, dport, sport, ip_proto)
+                    assert packet_port_index == rPacket_port_index
+
+                hit_count_map[packet_port_index] = hit_count_map.get(packet_port_index, 0) + 1
+            logging.info("outer_encap_fmts={}, hash_key={}, hit count map: {}".format(outer_encap_format, hash_key, hit_count_map))
+            if hash_key == 'outer-tuples':
+                self.check_all_packets_hash_to_same_nh(self.next_hop.get_next_hop(), hit_count_map)
+            else:
+                self.check_balancing(self.next_hop.get_next_hop(), hit_count_map)
+
+
+    def check_ip_route(self, hash_key, outer_encap_format, src_port, ip_src, ip_dst, sport, dport, ip_proto):
+        if (outer_encap_format == 'vxlan'):
+            (matched_index, received) = self.check_ip_route_vxlan(hash_key, src_port, ip_dst, ip_src, dport, sport, ip_proto)
+        elif (outer_encap_format == 'nvgre'):
+            (matched_index, received) = self.check_ip_route_nvgre(hash_key, src_port, ip_dst, ip_src, dport, sport, ip_proto)
+        else:
+            raise Exception("Logic issue, unexpected outer encap format {}".format(outer_encap_format))
+
+        matched_port = self.exp_port_list[matched_index]
+        return (matched_port, received)
+
+
+    def _get_ip_proto(self, ipv6=False):
+        # ip_proto 2 is IGMP, should not be forwarded by router
+        # ip_proto 254 is experimental
+        # MLNX ASIC can't forward ip_proto 254, BRCM is OK, skip for all for simplicity
+        skip_ports = [2, 254]
```

Comment:
@anish-n the variable name `skip_ports ` is confusing. Can we rename it accordingly (e.g, skip_proto)?

Reply:
Will do, thanks!

Reply:
Done, thanks